core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -2810,4 +2811,132 @@ class ReplicaManagerTest {
           Replicas.NONE, Replicas.NONE, 2, 123, 456)))),
     replicaManager.calculateDeltaChanges(TEST_DELTA))
   }
+
+  @Test
+  def testPartitionFetchStateUpdatesWithTopicIdAdded(): Unit = {
+    val aliveBrokersIds = Seq(0, 1)
+    val replicaManager = setupReplicaManagerWithMockedPurgatories(new MockTimer(time),
+      brokerId = 0, aliveBrokersIds)
+    try {
+      val tp = new TopicPartition(topic, 0)
+      val replicas = aliveBrokersIds.toList.map(Int.box).asJava
+
+      def leaderAndIsrRequest(epoch: Int, topicIds: util.Map[String, Uuid]): LeaderAndIsrRequest =
+        new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, 0, 0, brokerEpoch,
+          Seq(new LeaderAndIsrPartitionState()
+            .setTopicName(topic)
+            .setPartitionIndex(0)
+            .setControllerEpoch(0)
+            .setLeader(1)
+            .setLeaderEpoch(epoch)
+            .setIsr(replicas)
+            .setZkVersion(0)
+            .setReplicas(replicas)
+            .setIsNew(true)).asJava,
+          topicIds,
+          Set(new Node(0, "host1", 0), new Node(1, "host2", 1)).asJava).build()
+
+      val leaderAndIsrRequest1 = leaderAndIsrRequest(0, Collections.emptyMap())
+      val leaderAndIsrResponse = replicaManager.becomeLeaderOrFollower(0, leaderAndIsrRequest1, (_, _) => ())
+      assertEquals(Errors.NONE, leaderAndIsrResponse.error)
+
+      val fetchState = replicaManager.replicaFetcherManager.getFetcher(tp).flatMap(fetcher => fetcher.fetchState(tp))
+      assertTrue(fetchState.isDefined)
+      assertEquals(None, fetchState.get.topicId)
+
+      val leaderAndIsrRequest2 = leaderAndIsrRequest(0, topicIds.asJava)
+      val leaderAndIsrResponse2 = replicaManager.becomeLeaderOrFollower(0, leaderAndIsrRequest2, (_, _) => ())
+      assertEquals(Errors.NONE, leaderAndIsrResponse2.error)
+
+      val fetchState2 = replicaManager.replicaFetcherManager.getFetcher(tp).flatMap(fetcher => fetcher.fetchState(tp))
+      assertTrue(fetchState2.isDefined)
+      assertEquals(Some(topicId), fetchState2.get.topicId)
+
+    } finally {
+      replicaManager.shutdown(checkpointHW = false)
+    }
+  }
+
+  @Test
+  def testReplicaAlterLogDirsWithAndWithoutIds(): Unit = {
```

Comment:
Does this test work? I have the impression that we never update the `ReplicaAlterLogDirsManager`.

Reply:
You are correct that we don't update the IDs in ReplicaAlterLogDirsManager. 
This test runs an alter command with and without topic IDs (2 separate calls). I can make into two separate tests if that is clearer.

It does pass. I should rebase this pr so we can see the jenkins runs.

Reply:
Yeah, splitting the test is a good idea. It is pretty hard to read at the moment.

Reply:
With only 3 parameters changing based on two cases, I decided to make this a parameterized test instead. Hopefully it is clearer.